core/src/main/scala/org/broadinstitute/dsde/rawls/entities/EntityService.scala

```diff
@@ -339,35 +344,156 @@ class EntityService(protected val ctx: RawlsRequestContext,
       metadataFuture.recover(bigQueryRecover)
     }
 
-  def listEntities(workspaceName: WorkspaceName, entityType: String): Future[Seq[Entity]] =
-    getV2WorkspaceContextAndPermissions(workspaceName,
-                                        SamWorkspaceActions.read,
-                                        Some(WorkspaceAttributeSpecs(all = false))
-    ) flatMap { workspaceContext =>
-      dataSource.inTransaction { dataAccess =>
-        traceDBIOWithParent("countActiveEntitiesOfType", ctx) { countContext =>
-          dataAccess.entityQuery
-            .findActiveEntityByType(workspaceContext.workspaceIdAsUUID, entityType)
-            .length
-            .result
-            .flatMap { entityCount =>
-              if (entityCount > pageSizeLimit) {
-                throw new RawlsExceptionWithErrorReport(
-                  ErrorReport(
-                    StatusCodes.BadRequest,
-                    s"Result set size of $entityCount cannot exceed $pageSizeLimit. Use the paginated entityQuery API instead."
-                  )
-                )
-              } else {
-                traceDBIOWithParent("listActiveEntitiesOfType", countContext) { _ =>
-                  dataAccess.entityQuery.listActiveEntitiesOfType(workspaceContext, entityType)
-                }.map { r =>
-                  r.toSeq
-                }
-              }
+  /*
+   * Queries the db for a stream of entity attributes.
+   */
+  private def listEntitiesDbSource(workspaceContext: Workspace,
+                                   entityType: String
+  ): Source[EntityAndAttributesResult, NotUsed] = {
+    // note: ReadCommitted transaction isolation level; forward-only/read-only stream.
+    val allAttrsStream = dataSource.dataAccess.entityQuery
+      .streamActiveEntityAttributesOfType(workspaceContext, entityType)
+      .transactionally
+      .withTransactionIsolation(TransactionIsolation.ReadCommitted)
+      .withStatementParameters(rsType = ResultSetType.ForwardOnly,
+                               rsConcurrency = ResultSetConcurrency.ReadOnly,
+                               fetchSize = dataSource.dataAccess.fetchSize
+      )
+
+    // translate the Slick stream to a Source
+    Source.fromPublisher(dataSource.database.stream(allAttrsStream))
+  }
+
+  /**
+    * Given a Source containing entity attributes, scroll through that source and combine
+    * attributes into entities. Emit a Source of entities.
+    * <p>
+    * IMPORTANT: this !!requires!! that the incoming Source of attributes is ordered by
```

Comment:
Is there any way to enforce/check this ordering before any processing of entities?

Reply:
I struggled with this a whole bunch. The options I see are:
1. Inspect the entire result set before processing it to verify it is in the correct order. This defeats the entire purpose of streaming.
2. While iterating over the result set, detect if the current entity id is less than the previous entity id. Since we require that the result set is ordered by entity id, this would be an error case. However, since this is midstream we have already sent the `200 OK` status to the caller, and we could easily have already written some entities to the response payload. So, all we could do at this point is stop streaming after sending partial results, or potentially even encode an error response inside of a fake entity, a la`{"name": "ERROR", "entityType": "ERROR", "attributes": { "errorDescription" : "… error message here … "} }`.
3. Assume that this method is always called with an appropriate result set and enforce that via PR review and code inspection

I landed on the last, since option 2 has all kinds of weirdness about it. Do you have good suggestions on how to handle this? I will keep looking at other error-trapping options.

Reply:
Hmmm this is tough ha

To the overall goal, I think #3 is the best approach to your point -- just to clarify, what would an example incoming payload look like in terms of _what exactly_ would be ordered? Just want to make sure I understand this entirely

Reply:
The relevant code is the `order by e.id` here: https://github.com/broadinstitute/rawls/blob/6fd5c51f9fa44ed6133227d345d33042877f95b3/core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala#L340-L343.

I suppose that technically we don't actually care about _ordering_ - we just care that all attributes for any given entity are contiguous within the input stream. Remember that the input stream has one element per attribute, and the output stream has one element per entity. The implementation of our stream processing is that as soon as we see that the previous input element and the current input element belong to different entities, we output the previous entity and start collecting a new one. If attributes belonging to the same entity were not contiguous, we'd output that entity twice, and each would have only a portion of its attributes.

So, in pseudocode, this is a good input:

| entityId | entityName | attributeId | attributeName | attributeValue |
| --- | --- | --- | --- | --- |
| 1 | entity-one | 100 | color | red |
| 1 | entity-one | 101 | size | M |
| 2 | entity-two | 102 | color | green |
| 3 | entity-three | 103 | color | blue |
| 3 | entity-three | 104 | color | L |

which will produce the desired output of three entities:

```json
[
  {"name": "entity-one", "attributes": { "color": "red", "size": "M" }},
  {"name": "entity-two", "attributes": { "color": "green"}},
  {"name": "entity-three", "attributes": { "color": "blue", "size": "L" }}
]
```

but this is a bad input:

| entityId | entityName | attributeId | attributeName | attributeValue |
| --- | --- | --- | --- | --- |
| 1 | entity-one | 100 | color | red |
| 2 | entity-two | 102 | color | green |
| 3 | entity-three | 103 | color | blue |
| 1 | entity-one | 101 | size | M |
| 3 | entity-three | 104 | color | L |

which would produce undesirable output:
```json
[
  {"name": "entity-one", "attributes": { "color": "red" }},
  {"name": "entity-two", "attributes": { "color": "green"}},
  {"name": "entity-three", "attributes": { "color": "blue" }},
  {"name": "entity-one", "attributes": { "size": "M" }},
  {"name": "entity-three", "attributes": { "size": "L"}}
]
```

Reply:
Thank you for all of this 🙌 

Where my mind went to was wondering if we created a hash set alongside the streaming, where we would do a quick check of the hash set to see if an entity ID was encountered yet...but I assume that that would defeat the purpose of the streaming 🤔 

I think the `order_by` clause is probably enough then here? Let me know!

Reply:
I think the hashset approach would be essentially the same as option 2 above - we'd still be detecting an error midstream and have to figure out how to handle that. The more I think about it, though, we may want more error-handling for cases where something ELSE goes wrong, like losing the connection to the db. I'll try to handle that in the follow-on PR.

Reply:
Sounds good -- thanks for all the explanations here!